## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

 Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 0

If students address the detailed feedback in a future checkpoint they will earn these points back






|                                  | **Unsatisfactory**                                                                                                                                                                                                                                                                                                                        | **Developing**                                                                                                                                                                                                       | **Proficient**                                                                                                                                                                                            | **Excellent**                                                                                                                                                                            |
|----------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **EDA relevance**                | EDA is mostly neither relevant to the question nor helpful in figuring out how to address the question. Or the EDA does address the question, but many obviously relevant variables / analyses / figures were not included. | EDA is partly irrelevant/unhelpful. EDA missed one or two obvioulsy relevant analysis (distributions of single variables or relationships between variables) | EDA includes the obviously relevant / helpful variables in addressing the question.                                                              | Thorough EDA fully explored the dataset                                                                                                                 |
| **EDA analysis and description** | Many of the analyses are poor choices (e.g., using means instead of medians for obviously skewed data), or are poorly described in the text, or do not aid understanding the data                                                                                                                                                     | Some of the analyses are poor choices, or are poorly described in the text, or do not aid understanding the data                                                                                                 | All analyses are correct choices. Only one or two have minor issues in the text descriptions supporting them. Mostly they fit well with other elements of the EDA and support understanding the data  | All analyses are correct choices with clear text descriptions supporting them. The figures fit well with the other elements of the EDA, producing a clear understanding of the data. |
| **EDA figures**                  | Many of the figures are poor plot choices (e.g., using a bar plot to represent a time series where it would be better to use a line plot) or have poor aesthetics (including colormap, data point shape/color, axis labels, titles, annotations, text legibility) or do not aid understanding the data                                | Some of the figures are poor plot choices or have poor aesthetics. Some figures do not aid understanding the data                                                                                                | All figures are correct plot choices. Only one or two have minor questionable aesthetic choices. The figures mostly fit well with the other elements of the EDA and support understanding the data    | All figures are correct plot choices with beautiful aesthetics. The figures fit well with the other elements of the EDA, producing a clear understanding of the data.                |





# COGS 108 - EDA Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

# Research Question

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback



## Background and Prior Work

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

# Hypothesis


Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Data

### Data overview

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your data checkpoint feedback


In [ ]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [ ]:
# Setup code -- Run only once after cloning!!! 
#
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/airline-safety/airline-safety.csv', 'filename':'airline-safety.csv'},
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/bad-drivers/bad-drivers.csv', 'filename':'bad-drivers.csv'}
]

get_data.get_raw(datafiles,destination_directory='data/00-raw/')

### Dataset #1 

Instructions: REPLACE the contents of this cell and the one below with your work, including any updates to recover points lost in your data checkpoint feedback

In [ ]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION 

### Dataset #2
 as above, add any more copies of this that you need to given how many datasets you have

In [ ]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION 

### Dataset #3
### NBA Players Stats since 1950s from Kaggle
 
For Dataset #3, I am using a dataset from Kaggle called NBA player stats. This dataset contains
individual statistics for each player across 67 seasons. There are many useful column variables to
work with for measuring a player’s career success, such as win shares, win shares per 48 games,
true shooting percentage, and more. However, since this dataset contains players from before the 2000s and
many of these statistics didn’t exist before then, we need to consider filtering these players out of
our data (dataset #2 also only contains players beyond 2000s, so we should filter out players before then). Additionally, since the dataset contains repeated entries of players due to it capturing the stats of players for every season, we need to sum up or average the stats of interest for their total career statistics. 

There many column variables in this dataset, including win shares, true shooting percentage, age, block percentage, free throw rate, etc. However, when measuring a player's career success, there are a few columns that can be major indicators. The columns we are most interested in are win shares, true shooting percentage, total career points, total games played, and total career minutes. 

The winshares are a cumulative statistic that estimates the exact number of team wins contributed by a player due to both their offense and defense. Essentially, they represent the overall impact on team success made by a player. The true shooting percentage measures how effectively a player scores, accounting for 2-pointers, 3-pointers and free throws in a single metric. This statistic represents the offensive efficiency of a player. The total points are the total points scored by the player throughout their career, which represents their raw offensive output. The total games represent the longetivity and durability by a player because it measures how long a player remained healthy throughout their career. The total minutes is how much time the player spent actually playing. which is why it represents the level of trust the player's coach has with the player and how big of a role they played for their teams. When we combine these statistics together, we are evaluating the overall success impact, offensive efficiency, volume of success, reliability, and longevity of a player, which effectively captures the player's career success. 

### A: Loading Data Set


In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt
import os
# load both career and draft data to prepare for merge
raw_path = 'data/00-raw/draft-data-20-years.csv'
df_combine = pd.read_csv(raw_path)
df_career = pd.read_csv('data/00-raw/Seasons_Stats.csv')
# merge into dataframe with all players with both combine and career data
df_final = pd.merge(df_combine, df_career, on='Player', how='inner')

### B: Make the dataset tidy or demonstrate that it was already tidy

In [2]:
 # create a new dataframe with ONLY the columns we are interested in
my_columns = [
'Player', 'Draft_Year',
'G_y', 'MP', 'WS_y', 'TS%', 'PTS']
existing_cols = [c for c in my_columns if c in df_final.columns]
df_project1 = df_final[existing_cols].copy()
df_project1 = df_project1.rename(columns={
'G_y': 'Games',
'WS_y': 'Win_Shares',
'MP': 'Minutes',
'PTS': 'Points',
'TS%': 'True_Shooting'
})

In [3]:
 # Sum the volume stats and average the true shooting percentage
aggregations = {
'Games': 'sum',
'Minutes': 'sum',
'Win_Shares': 'sum',
'Points': 'sum',
'True_Shooting': 'mean'
}
# since this dataset shows each player's season stats, we want to sum all of their seasons stats into one single data for each player
df_project = df_project1.groupby('Player').agg(aggregations).reset_index()
df_project.head(10)

,Player,Games,Minutes,Win_Shares,Points,True_Shooting
0,A.J. Bramlett,8.0,61.0,-0.2,8.0,0.190000
1,A.J. English,151.0,3108.0,1.1,1502.0,0.480000
2,A.J. Guyton,80.0,1246.0,0.3,442.0,0.324000
3,A.J. Hammons,22.0,163.0,0.0,48.0,0.472000
4,A.J. Price,287.0,4253.0,5.0,1656.0,0.435000
5,Aaron Brooks,797.0,17088.0,23.2,7839.0,0.512267
6,Aaron Gordon,205.0,4958.0,10.1,1981.0,0.529333
7,Aaron Gray,387.0,4517.0,7.1,1235.0,0.525000
8,Aaron McKie,957.0,22594.0,44.9,6636.0,0.493059
9,Acie Earl,240.0,3012.0,0.3,1168.0,0.444333


### C: Demonstrate the size of the dataset

In [4]:
print(f"Dataset Dimensions: {df_project.shape}")
print(f"Rows (Players): {df_project.shape[0]}")
print(f"Columns (Variables): {df_project.shape[1]}")

Dataset Dimensions: (1221, 6)
Rows (Players): 1221
Columns (Variables): 6


### D: Investigate Missing Data

In [6]:
# Count missing values per column
missing_counts = df_project.isnull().sum()
print("\n--- Missing Data Analysis---")
print(missing_counts[missing_counts > 0]) # Only print columns with gaps

# Calculate percentage of missing data
print((missing_counts / len(df_project)) * 100)


--- Missing Data Analysis---
True_Shooting    3
dtype: int64
Player           0.0000
Games            0.0000
Minutes          0.0000
Win_Shares       0.0000
Points           0.0000
True_Shooting    0.2457
dtype: float64


### E: Detect Outliers or Suspicious Entries

In [8]:
numeric_cols = ['Games', 'Minutes', 'Win_Shares', 'Points', 'True_Shooting']
print("---Outlier Detection Report---")
# detect outlier using the interquartile method
for col in numeric_cols:
    
# calculate quartiles
    Q1 = df_project[col].quantile(0.25)
    Q3 = df_project[col].quantile(0.75)
    IQR = Q3- Q1
    lower_bound = Q1- 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

# Find the outliers
outliers = df_project[(df_project[col] < lower_bound) | (df_project[col] > upper_bound)]

# Print findings
print(f"\n Variable: {col}")
print(f"- Normal Range: {lower_bound:.2f} to {upper_bound:.2f}")
print(f"- Outliers Found: {len(outliers)}")

# Show the top 3 biggest outliers
if not outliers.empty:
    print(f"- Top 3 Outliers:")
    print(outliers[['Player', col]].sort_values(by=col, ascending=False).head(3).to_string(index=False))
    print("\n-----------------------------")
    
suspicious_negatives = ['Games', 'Minutes', 'Points'] #cannot be negative here, (note that win shares can be negtative if a player performs bad enough)
print("--- Suspicious Value Check (Negative Values)---\n")

# Check for the strictly positive columns
for col in suspicious_negatives:
    # Filter for values less than 0
    negative_values = df_project[df_project[col] < 0]
if len(negative_values) > 0:
    print(f"\nFound {len(negative_values)} negative values in '{col}'")
    print(negative_values[['Player', col]])
else:
    print(f"'{col}': No negative values found. (Clean)\n")
    print("------------------------------------------")

---Outlier Detection Report---

 Variable: True_Shooting
- Normal Range: 0.35 to 0.64
- Outliers Found: 73
- Top 3 Outliers:
        Player  True_Shooting
 Tyson Wheeler          1.064
Chad Gallagher          1.000
 Randy Holcomb          1.000

-----------------------------
--- Suspicious Value Check (Negative Values)---

'Points': No negative values found. (Clean)

------------------------------------------


### F: Clean the Data

In [12]:
# Fill missingnumeric stats with 0
cols_to_fix= ['Games', 'Minutes', 'Win_Shares','Points', 'True_Shooting']
for col in cols_to_fix:
    df_project[col]= df_project[col].fillna(0)

# Verify cleaning
print(f"\nMissing values after cleaning: {df_project.isnull().sum().sum()}")


Missing values after cleaning: 0


### G. Save Cleaned Dataset

In [13]:
processed_path_career= 'data/02-processed/dataset_3_career_clean.csv'

df_project.to_csv(processed_path_career,index=False)

print("Cleaned dataset saved.")

Cleaned dataset saved.


### Summary Statistics of Dataset #3

In [14]:
summary = df_project.describe().round(2)

print("\n--- Summary Statistics (Career Data)---")
print(summary)
print("\n")


--- Summary Statistics (Career Data)---
         Games   Minutes  Win_Shares    Points  True_Shooting
count  1221.00   1221.00     1221.00   1221.00        1221.00
mean    388.89   9280.51       19.03   3828.40           0.49
std     353.64  10584.35       28.45   4991.04           0.09
min       1.00      0.00       -2.40      0.00           0.00
25%      80.00    931.00        0.50    285.00           0.46
50%     279.00   5144.00        6.40   1801.00           0.50
75%     628.00  14742.00       26.90   5728.00           0.53
max    1884.00  54981.00      206.30  33643.00           1.06




What this means:

These summary statistics show a right-skewed distribiution in player performance, where the “average” entry represents a player who plays around 389 games and scores a total of 3828 points in their career, where as the 50% median is lower at around 1801 points. This data also shows that there is a major gap between typical players and elite players, where the top 25% of players generate an impressive 206.30 winshares and 33,643 total points, which is disproportionately greater than the rest. This is the reason why the mean is noticeably greater than the median; the elite players are skewing the avergae up. The maximum values, such as 33,643 points, align with historical NBA records which confirms the data cleaning and tidying process.

## Results

### Exploratory Data Analysis

Instructions: replace the words in this subsection with whatever words you need to setup and preview the EDA you're going to do.   

Please explicitly load the fully wrangled data you will use from `data/02-processed`.  This is a good idea rather than forcing people to re-run the data getting / wrangling cells above.  Sometimes it takes a long time to get / wrangle data compared to reloading the fixed up dataset.

Carry out whatever EDA you need to for your project in the code cells below.  Because every project will be different we can't really give you much of a template at this point. But please make sure you describe the what and why in text here as well as providing interpretation of results and context.

Please note that you should consider the use of python modules in your work.  Any code which gets called repeatedly should be modularized. So if you run the same pre-processing, analysis or visualiazation on different subsets of the data, then you should turn that into a function or class.  Put that function or class in a .py file that lives in `modules/`.  Import the module you made and use it to get your work done.  For reference see `get_raw()` which is inside `modules/get_data.py`. 



#### Section 1 of EDA - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [ ]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

#### Section 2 of EDA if you need it  - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [ ]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

## Ethics

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Team Expectations 

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Project Timeline Proposal

Instructions: Replace this with your timeline.  **PLEASE UPDATE your Timeline!** No battle plan survives contact with the enemy, so make sure we understand how your plans have changed.  Also if you have lost points on the previous checkpoint fix them